iwyu_ast_util.cc

```diff
@@ -549,6 +549,11 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
   // --- Methods on RecursiveASTVisitor
   bool VisitType(Type* type) {
     seen_types_.insert(type);
+
+    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
```

Comment:
Hmm. `getAs` is a little deceptive here. If I recall correctly, it shaves off sugar until it reaches `TemplateSpecializationType`. So if you have a `TemplateSpecializationType` nested inside several layers of sugar, e.g. `ElaboratedType`, `TypedefType`, I think you might end up traversing the template argument multiple times.

For example (off the top of my head, might not be 100% in line with reality):
```
// Input
typedef class SomeTemplate<int> SomeSpecialization;

// AST:
// TypedefType: typedef class SomeTemplate<int> SomeSpecialization
// ElaboratedType: class SomeTemplate<int>
// TemplateSpecializationType: SomeTemplate<int>
```
In this case, it seems to me `VisitType` would see `TypedefType`, then `ElaboratedType`, then `TemplateSpecializationType`, and `getAs` will return the same node for all three.

Any chance you can provoke something like that and see if it actually happens?

So while `getAs` is often the right thing to do, if the above happens, I think `dyn_cast` is a better alternative, because we only want to trigger for the particular node that _is_ a `TemplateSpecializationType`. But I could definitely be wrong.

Reply:
The original problem is that `RecursiveASTVisitor` doesn't traverse, e.g., typedef underlying type by default. This is why I had to write these lines of code. For the case of explicitly written template specialization type, it already worked fine and traversed its arguments. Hence, some redundancy actually present here, but I don't think it is critical, because `seen_types` is a `set`, and duplications are automatically removed.

Reply:
Ah, so `RecursiveASTVisitor` doesn't reach the `TemplateSpecializationType` from a `TypedefType` or a `UsingType`? Let's add a comment before the `getAs` call to clarify that. You _could_ probably override `TraverseTypedefType` and `TraverseUsingType` to also traverse the underlying type instead, but it seems like a weaker change (TST under other kinds of sugar wouldn't be handled).

One thing I would like, though, is to avoid `Visit` calling `Traverse`. Usually traversal tweaking happens in `TraverseX`, whereas the thing you want to do with the nodes should go in `VisitX`. There's some rationale in this comment: https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/RecursiveASTVisitor.h#L87, particularly this part: https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/RecursiveASTVisitor.h#L124.

So I would suggest a new `TraverseType` with the special traversal rule, and a comment to explain that it's necessary because some sugar nodes (I've found `TypedefType` and `UsingType`, but there may be more) do not visit the underlying type, and hence the template arguments would never be seen.